### ISSUE TITLE
printf format overflows

### ISSUE THREAD

**Issue by loveshack** [flame/blis#178]
gcc 7 warns about format overflows, with this note

````
./include/knl/blis.h:27445:3: note: ‘sprintf’ output 8 bytes into a destination\
 of size 7
   sprintf( func_str, "%s%-5s", dt_str, op_str ); \
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
````

This fixes it:

````
diff --git a/frame/compat/bli_blas.h b/frame/compat/bli_blas.h                  
index 242a03b..3a732ef 100644                                                   
--- a/frame/compat/bli_blas.h                                                   
+++ b/frame/compat/bli_blas.h                                                   
@@ -52,7 +52,7 @@
                                                                                
 // -- Constants --                                                             
                                                                                
-#define BLIS_MAX_BLAS_FUNC_STR_LENGTH (6+1)                                    
+#define BLIS_MAX_BLAS_FUNC_STR_LENGTH (7+1)                                    
                                                                                
                                                                                
 // -- Utility macros --                                                        
````

**Comment by fgvanzee** [comment]
Thanks Dave, your fix has been applied in 3d1a5a7.

I can't really convince myself of *why* the string was overflowing, though. `dt_str` is a single character string (excluding `\0` terminator). `op_str` is at most five characters (again, excluding the terminator). So I can count 1+5+1 characters (including terminator) going into `func_str`. What am I missing?

**Comment by devinamatthews** [comment]
@fgvanzee never ever use `sprintf`. Use `snprintf` or `asprintf` instead (the latter is a GNU extension but is very easy to implement yourself in terms of two calls to the former).

**Comment by fgvanzee** [comment]
@devinamatthews Could you explain? It's not clear to me what is wrong with `sprintf()`.

**Comment by devinamatthews** [comment]
`sprintf` == possible (in practice probable) buffer overflow. Of course, I would just use C++:

```C++
string dt_str = ...;
string op_str = ...;
auto func_str = dt_str + op_str;
```

but that's neither here nor there.

**Comment by fgvanzee** [comment]
So you're just suggesting to never use `sprintf()` to avoid shooting myself in the foot? I thought you were going to say that it was inherently dangerous even when used properly. (It's still not clear to me what went wrong with the original code.)

**Comment by jeffhammond** [comment]
@fgvanzee The fact that you don't know what went wrong in the original code suggests that proper use of this function is harder than it looks 😉

**Comment by fgvanzee** [comment]
@jeffhammond Allow me to clarify. I'm not trying to argue that `snprintf()` isn't safer than `sprintf()`. I'm fine with migrating to the former. But when I "make a mistake" I prefer to understand why it's a mistake. Specifically, in this case, I would like to understand how my strings ended up being longer than seven characters (including terminator). *Or*, maybe I didn't do anything wrong and this is a compiler bug. Either way, I prefer to know what's going on.

**Comment by loveshack** [comment]
You wrote: 

> Thanks Dave, your fix has been applied in 3d1a5a7.
>
> I can't really convince myself of *why* the string was overflowing,
> though. `dt_str` is a single character string (excluding `\0`
> terminator). `op_str` is at most five characters (again, excluding the
> terminator). So I can count 1+5+1 characters (including terminator)
> going into `func_str`. What am I missing?

I looked at the pre-processed output (cc -E) of the original, which
contains

char func_str[ (6+1) ]; sprintf( func_str, "%s%-5s", "ch", "ger" );

so the buffer length needs to be 2+5+1.

[sprintf is one of the first things I'd check on for programming.  While
using snprintf saves you overflowing the buffer, of course you need to
check the return value if truncation would matter.  I've seen BLAS as a
dependency of some system-type software I wouldn't have expected, so I
guess we should take care even with numerical libraries; Pabst has
recently put some hardening flags in the libxsmm build.]

**Comment by fgvanzee** [comment]
> char func_str[ (6+1) ]; sprintf( func_str, "%s%-5s", "ch", "ger" );
>
> so the buffer length needs to be 2+5+1.

Thanks Dave. Your quoting of cpp output got me looking into the code closer. Before, I always looked at another operation's file, like `bla_gemm.c`, because I figured they all suffered from the same thing. But then I noticed you were specifically citing `ger` as the offender. That set me on the right path and eventually I found the problem:
```
#undef  GENTFUNCDOT
#define GENTFUNCDOT( ftype, chxy, chc, blis_conjy, blasname, blisname ) \
\
void PASTEF772(chxy,blasname,chc) \
     ( \
       const f77_int* m, \
       const f77_int* n, \
       const ftype*   alpha, \
       const ftype*   x, const f77_int* incx, \
       const ftype*   y, const f77_int* incy, \
             ftype*   a, const f77_int* lda  \
     ) \
{ \
    ...
    /* Perform BLAS parameter checking. */ \
    PASTEBLACHK(blasname) \
    ( \
      MKSTR(ch), \
      MKSTR(blasname), \
      m, \
      n, \
      incx, \
      incy, \
      lda  \
    ); \
    ...
```
I was using a macro parameter of `chxy` for the character encoding the datatype of the operation, but then I was using `ch` in the call to the `_check()` function (or, `_check()` macro as of c73055f). Since there was no `ch` macro parameter, it was not substituted with anything prior to it being string-ified via `MKSTR()`, hence the length "2" (excluding terminator) for the datatype string. I've applied a fix in fe7d7f1. Thanks again for your help with this.

**Comment by fgvanzee** [comment]
I think this is all fixed now. Closing.
